rfd/xxxx-config-resource-docs.md

```diff
@@ -0,0 +1,554 @@
+and so on). The sequence is:
+
+- Look up a user-defined struct type from the configuration within the
+  declaration map.
+- Construct a `Resource` from the struct type and insert it into the final data
+  map (unless a `Resource` already exists there).
+- Process the fields of the user-defined struct type using the rules described
+  [below](#processing-field-types).
+
+Once this is done, the generator will convert the  `map[declKey]Resource` into a
+single `[]Resource` to feed into the template.
+
+#### Alternative: Using analysis drivers
+
+The Go Tools repository provides the `multichecker`
+(https://pkg.go.dev/golang.org/x/tools/go/analysis/multichecker) and
+`singlechecker`
+(https://pkg.go.dev/golang.org/x/tools/go/analysis/singlechecker) packages
+within `golang.org/x/tools/go/analysis`, which run custom static analysis
+functions, `analysis.Analyzer`s. However, these static analysis tools are
+intended for linting, not generating reference documentation. 
+
+While you can configure an `analysis.Analyzer` to return a value from its `Run`
+function, the `multichecker` and `singlechecker` drivers run by exposing a
+`Main` function that handles all of the program's output and CLI arguments. It
+is not possible to initialize a data structure outside the `Main` function and
+assign values to it during the `Run` function, since `Main`
+[exits](https://cs.opensource.google/go/x/tools/+/refs/tags/v0.8.0:go/analysis/multichecker/multichecker.go;l=59)
+the current process after completing, leaving the program no chance to process
+the data structure.
+
+The `singlechecker` and `multichecker` packages run static analysis by importing
+from a `checker` package, which is internal to the
+`analysis` package. While we could copy this code and
+modify it for our purposes, the reference generator does not require the
+general-purpose static analysis capabilities of the `analysis` package.
+
+### Converting structs to `Resource`s
+
+In Go's `ast` package, a [`Decl`](https://pkg.go.dev/go/ast#Decl) is an
+interface representing a declaration. The logic for generating a `Resource`
+from a given type declaration depends on the nature of the type declaration. 
+
+Start by asserting that the `Decl` is an `*ast.GenDecl` with a `Specs`
+containing a `TypeSpec` with `Type: StructType`. From there, we can map the
+fields within [`ast.StructType`](https://pkg.go.dev/go/ast#StructType) to fields
+in a new `Resource`.
+
+Each `StructType` has a [`*FieldList`](https://pkg.go.dev/go/ast#FieldList) that
+we can use to obtain field data, including comments and tags.
+
+|`Resource` Field|How to generate|
+|---|---|
+|`Resource.SourcePath`|The `file` property of the underlying `sourceDecl`.|
+|`Resource.Description`|`GenDecl.Doc`|
+|`Resource.SectionName`|`TypeSpec.Name.Name`, with camelcase converted to separate words.|
+|`Fields[*].Name`|Process the `Field.Tag` in the `FieldList`, extracting the value of the `json` tag.|
```

Comment:
Do we expect json tags to always exist?

Reply:
Ah, that's a good point. Should the generator check whether the struct type has an `UnmarshalYAML` or `UnmarshalJSON` method before processing the struct declaration AST? 

If there is an `UnmarshalYAML` or `UnmarshalJSON` method, the generator would treat the struct as a custom type, since it won't be able to tell how struct field names translate into YAML field names. If there are no such methods, the generator would use the original struct field name.

Reply:
I'm more interested in what our "requirements" are - it's alright if we want all structs to have JSON tags, as long as we say so here.

Special handling of custom unmarshal methods seems like a good idea, or at the very least some warnings from the generator.